### PR TITLE
Allow custom parsing of EntitySelector [1.11]

### PR DESCRIPTION
**Pull request by Mathe172**
1.11 version of #3353

One minor change: `SelectorHandler.matchEntities(...)` and `.matchesMultiplePlayers(...)` now throw `CommandException` since Vanilla now does.

### CODE REVIEW

File: src/main/java/net/minecraftforge/common/command/SelectorHandler.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    public static void setHandler(SelectorHandler handler)
+    {
+        if (isCustom)
+            FMLLog.bigWarning("SelectorHandler is already overridden" + (modifyingMod == null ? "" : " by '" + modifyingMod + "'"));
+
+        isCustom = true;
+        ModContainer container = Loader.instance().activeModContainer();
+
+        if (container == null)
+            FMLLog.bigWarning("SelectorHandler is overridden by an unknown mod. Overriding should be done during mod loading");
```
Comment (mezz) [inline src/main/java/net/minecraftforge/common/command/SelectorHandler.java]: This one should throw an exception if it's important to know the modifying mod.

File: src/main/java/net/minecraftforge/common/command/SelectorHandler.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+import java.util.List;
+
+/**
+ * A proxy for {@link EntitySelector} methods. Can be used to customize the parsing behavior<br>
+ * <b>Note:</b>  You should generally override all three methods to produce consistent results<br>
+ * <b>Note:</b> These methods are called on both client and server<br>
+ * <b>Note:</b>  For compatibility reasons, you should not break the key=value pattern
+ */
+public class SelectorHandler
```
Comment (mezz) [inline src/main/java/net/minecraftforge/common/command/SelectorHandler.java:35]: It might be better if you split this class into two like `SelectorHandler` and `SelctorHandlerManager`, so all the static stuff is separated out.

**Review by mezz — CHANGES_REQUESTED** [review]
This is a bit of a strange PR because it's very single-use. Only one mod can change the commands at a time, since there can only be one `SelectorHandler`. I guess that's probably fine since it's unlikely that other mods will want to do what you're doing, and there may not be a sane way to combine handlers.

**Comment by PhiProven** [comment]
Updated the PR
`SelectorHandlerManager.setHandler(...)` now throws an `NPE` if `Loader.instance().activeModContainer()` returns `null`. This should probably never happen and therefore just be used to determine which mods screws up by calling this method outside mod loading.

**Comment by PhiProven** [comment]
We know that the PR is single-use, but we couldn't think of any other way to accomplish this other than a coremod (or by overriding every single instance of `CommandResultStats` and `ITextComponent` which would be much more awkward). Since Lex does not like coremods, we wanted to try it with this PR first.

One possible way of combining handlers is probably by using the previous one as fallback if the new one fails. But I think that it is best if the new handler does that by itself (it can query the previously set handler before overriding it and then use it as fallback)

**Review by mezz — APPROVED** [review]

**Comment by mezz** [comment]
Yeah that makes sense, looks good to me.

**Comment by mezz** [comment]
After talking to Lex about this, it would make a lot more sense as a direct change to Forge.
Since only one mod can change the selector handler, it would be best to simply have a good agreed-upon format extension in Forge rather than create a hook here and put the implementation in a mod.

Can you make a new PR for the parsing format and host some discussion about it there?

**Comment by Mathe172** [comment]
To give a bit of background on our situation (well, a huge wall of text...): We currently have (or, to be more precise, we're in the process of writing) a very general framework for command and selector parsing, which we would very much like to keep (in the sense that we don't want to sacrifice any features). A quick overview over what it does:

* Commands are parsed into tree structures representing their evaluation, making it possible to execute it multiple times without reparsing it. By representing the sub parts as objects rather than plain strings and implementing a type system handling the conversion between different types, it is possible to just insert almost any sub expression almost anywhere it makes sense. (Note that the type system is almost invisible to the end user)
* Instead of parsing everything by themselves, the commands declare what type of data they want (e.g. a position, a list of entities or a NBT tag). This is done by providing a parser for that argument (obviously, there are parsers for all standard types)
* By handling the parsing through parsers, the command syntax can easily be upgraded, e.g. by adding new selector like structures for every type and by adding a way to capture evaluation results to use them later on. The aforementioned type system makes this even more powerful, since these selector like expressions can be used almost anywhere (even nested inside each other)
* Arguments can be anything (not just space delimited tokens). A parser for a specific argument just gets the current position in the string to parse, and can increment that position however it chooses.
* The parsing is managed by a central Parser that can be used to exchange data between the individual sub parsers
* Every sub parser can declare metadata such as completion for itself. This makes it possible to e.g. provide accurate completion everywhere for every part of the command (for the case of NBT compounds, the completion even knows the structure of the compound and can complete only appropriate keys)

The issue now is that we don't see any way of creating a simple framework which would allow us to embed ours (apart from just handing over the entire parsing to the mod, as this PR would do it). The only other option would be to more or less implement the whole framework into Forge itself. While we would happily provide the framework, we see a few issues with that:

* The framework is most likely too overengineered to be provided by Forge itself (it's quite a lot of code for something that's not the core purpose of Forge)
* The whole framework consists of ~150 files. While the core alone is only ~50, it would be more or less useless (since this would basically mean you'd have to implement many things by yourself) without adding the rest with a mod (which everyone using the framework would have to depend on)
* While the vanilla commands would work through the legacy support, they would have to be completely overridden to benefit from the new syntax (as the ICommand interface simply doesn't provide enough information). Of course, this would mean that every command would have to be updated every time Minecraft is updated.

*tl;dr;* We have a framework, and we would like to keep it. The only way to allow it we see is by transferring the whole control over the parsing to the mod (i.e. this PR), or by adding the framework itself to Forge.
If you'd still prefer the whole framework, we would need a while to clean it up and finish with all the API breaking stuff we've planned. (Please not the GitHub project is horribly out of date - the reason is that the project structure is very broken, as we unfortunately didn't start as a forge mod. If you'd like, we could upload the most important parts)

**Comment by mezz** [comment]
Thanks for the info.
That's definitely on a much larger scale than I imagined.  I'll talk to Lex again when he's online and see what he thinks about it.

**Comment by LexManos** [comment]
With that description and what your full intentions being a full re-write of the entire command system.
I'm going to deny this PR and state that any such special case massive overhaul of the command system like you're speaking of should be done as its own standalone mod.
The current MC system for commands allows you to replace the command manager at the lowest level. 
{Requires a little reflection but you're hacking things anyways}
There appears to be no method of providing this hook that would even come close to allows mods to play nice together.

What I would suggest to you is to replace the command manager.
Abandon ICommand and re-write every command that is in vanilla to use your system to its fullest.
When it comes to modded commands you will have to provide compatibility modules for your mod.
This will be on you to provide. 

So ya, as far as this goes its way out of scope of Forge. And there are ways to take full control and do what you wan't already. WITHOUT A COREMOD.

**Comment by Mathe172** [comment]
We know that this is possible for the CommandHandler (and it's what we intend to do). However, we see no way to achieve similar things for selectors. While we can modify selectors in commands, there is, as pointed out in the opening post of the original PR, no way to modify them globally (or at least we don't see any feasible one). As mentioned, we know that you could theoretically hack into the CommandResultStats and ITextComponent using a ton of reflection (essentially replacing every single instance) - but this is arbitrarily messy and incompatible.
The point of this PR was therefore to add an equivalent to the CommandHandler for selectors (hence the name) to allow modifying every selector.

**Comment by LexManos** [comment]
The command handler is the entry point.
It is where the core "do something with this string" is.
If you redirect that method nothing should ever call the vanilla selectors.
I will not walk you through this. Good luck, it'll be a lot of work.

**Comment by Mathe172** [comment]
The point is that the `CommandHandler` is **not** a common entry point:
`ITextComponents` and `CommandResultStats` both call `EntitySelector.matchEntities` without going through `CommandHandler` (although there is a `CommandHandler` somewhere in the call stack most of the time).

As an example for a case where there is no `CommandHandler` at all in the call stack: The following stack trace was created by setting the first line of text of a sign to `{"selector":"@e"}` and loading the world:
```
net.minecraft.command.EntitySelector.matchEntities(EntitySelector.java:119)
net.minecraft.command.EntitySelector.matchEntitiesToTextComponent(EntitySelector.java:98)
net.minecraft.util.text.TextComponentUtils.processComponent(TextComponentUtils.java:51)
net.minecraft.tileentity.TileEntitySign.readFromNBT(TileEntitySign.java:107)
net.minecraft.tileentity.TileEntity.create(TileEntity.java:121)
net.minecraft.world.chunk.storage.AnvilChunkLoader.loadEntities(AnvilChunkLoader.java:489)
net.minecraftforge.common.chunkio.ChunkIOProvider.syncCallback(ChunkIOProvider.java:96)
net.minecraftforge.common.chunkio.ChunkIOExecutor.syncChunkLoad(ChunkIOExecutor.java:94)
net.minecraft.world.gen.ChunkProviderServer.loadChunk(ChunkProviderServer.java:119)
net.minecraft.world.gen.ChunkProviderServer.loadChunk(ChunkProviderServer.java:90)
net.minecraft.world.gen.ChunkProviderServer.provideChunk(ChunkProviderServer.java:136)
net.minecraft.server.MinecraftServer.initialWorldChunkLoad(MinecraftServer.java:302)
net.minecraft.server.integrated.IntegratedServer.loadAllWorlds(IntegratedServer.java:106)
net.minecraft.server.integrated.IntegratedServer.startServer(IntegratedServer.java:120)
net.minecraft.server.MinecraftServer.run(MinecraftServer.java:442)
java.lang.Thread.run(Thread.java:745)
```
This is just to provide an example where Vanilla (and Forge without mods) call `EntitySelector` methods without going through the `CommandHandler` (this particular case could probably be fixed by modifying `TileEntitySign`, but this is again a fix that can't possibly cover all cases)

**Comment by PhiProven** [comment]
To extend Mathe172's answer:
`ITextComponent` and `CommandResultStat` internally call `EntitySelector.matchEntities()` to parse the selectors. Unfortunatetely, there is no feasible Vanilla solution to reroute those calls over our code.
We would have to write our own version of the entire classes just to apply this small patch. As they are not affected otherwise by our upgraded command parsing, this would be an overkill as it is really messy and a good source for compatibility issues. Additionally, this would only cover those two special cases but would not work nicely with any other mod that calls `EntitySelector.matchEntities()`. (We could alternatively wrap every single instance of those classes to override the appropriate methods, but that would give rise to the same problems.) 
Rerouting `EntitySelector.matchEntities()` itself should in contrast be very compatible with other mods and much easier to maintain.

**Comment by PhiProven** [comment]
Any update on this?
(We know how to override `CommandHandler` to implement our custom command system and don't ask for help about that. That wasn't the point of this PR anyway.)
Our problem remains: Though most selectors are evaluated during command execution, we don't have direct control over that. It is done internally by some Vanilla classes and, for obvious compatibility and maintainance reasons, we don't want to reinvent them just to patch the selector parsing. This PR would allow us to override the selector syntax in a safe way. Do you have any other suggestions how to achieve the same in a compatible way?

**Comment by mezz** [comment]
After talking more with Lex, unfortunately this PR will not be accepted.

Full replacement of vanilla features is generally not a perfectly compatible thing to do, and this type of patch is just past the edge of what can be considered part of the "compatibility layer" which defines Forge.

There are ways to implement this as you have stated, for instance it may make sense to implement your own versions of the commands that make use of your selector, while leaving in the vanilla commands alone for compatibility. You may find that you want or need this level of control for other features anyway.

I'm sorry but this PR does not fit with Forge, thankfully it should not be required for your project to succeed.

**Comment by PhiProven** [comment]
The Vanilla commands are not the problem. We planned to write our own version of them anyway.
The Vanilla classes I was talking about are `ITextComponent` and `CommandResultStats`. Writing our own version of them is worse than rewriting a command, as they are used widely across the code while commands don't interact with each other.
Do you have any suggestion how to replace them in a safe and maintainable way? (Here I'm indeed asking for a little help, as we don't see a nice way of doing that, apart from reinventing the whole classes which is not really what we want...) :)

**Comment by PhiProven** [comment]
After some further consideration, I've come to the conclusion that it is not possible to write our own version of `CommandResultStats` (at least without strange effects):
As Vanilla commands (and any commands added by other mods) would still use the old `CommandResultStats`, by writing an independent implementation for them, we would force commandblockers to use two different versions of them (depending on whether the executed command is Vanilla or not). This is really undesireable and in no way backwards compatible, so writing an independent implementation is no option.
Another option would be to extend `CommandResultStats` and override the appropriate methods. The problem with this is that we had to replace every instance of them in order to get consistent behavior. Unfortunately, they are created simply by `new` and not some `create()` method. This makes it impossible to add a general Event for replacing them upon construction. We *could* add a PR that fires an event (which would allow to replace the generated instance of `CommandResultStats`) everywhere Vanilla generates them. But this would probably be even more awkward than this PR and would screw with any mod that generates them on its own.

Btw, thanks for taking your time on that :)

**Comment by PhiProven** [comment]
Would the following little change to this PR help?
Instead of having a single `SelectorHandler` in charge of everything, we could have multiple ones associated to "domains" (where "domain" here just means prefix of the "selector" string). For example, Vanilla would have `SelectorHandler`s registered for the domains `@p, @a, @r, @e`. If you wanted to add another selector `@s` (for selecting the sender itself), you could do that by registering a handler for the domain `@s`.
We could then register a handler for our upgraded selectors for the domain `#@` or something. I prefer to write `#@e` or something for the upgraded version of `@e` instead of somehting like `@ue` (upgraded e), hence the more general notion of "domains" instead of just selector names. This may come in handy anyway.
This way, mods work nicely together, as long as they don't use the same selector name (or domain) (I don't want to enforce a naming convention since nobody wants to write unnecessary long selector names in commands). This would also not override the Vanilla selectors but rather provide the possibility to add an upgraded version with a different name (`#@`or something).
I'm not sure, how conflicting domains should be handled (e.g. you have domains `@` and `@s` and type in something like `@s[...`). I'd suggest to use the best match (i.e. `@s`) and just generate a warning at registration.

Though something like this is already possible with the current state of the PR (as a `SelectorHandler` can use the previously installed one as fallback), this approach will move control to Forge which is probably safer.

**Comment by mezz** [comment]
That sounds like a more compatible solution. I'll run it by Lex and see what he thinks, he's looked into this a lot further than I have.

**Comment by Mathe172** [comment]
Should we create a new PR/update this one with the new proposal to make it clearer what we mean, or is @PhiPro95's description good enough?

**Comment by mezz** [comment]
Code almost always helps clarify, and often things change slightly in implementation, so that would be helpful.
